File: src/main/java/gg/uhc/migration/MigrationRunnable.java
package gg.uhc.migration;

import com.google.common.collect.Maps;
import gg.uhc.migration.messages.MessageSender;
import gg.uhc.migration.selection.AreaSelection;
import org.bukkit.Bukkit;
import org.bukkit.GameMode;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.List;
import java.util.Map;
import java.util.UUID;

public class MigrationRunnable extends BukkitRunnable {

    enum Phase {
        WARNING,
        DAMAGING
    }

    // weighted collection of all areas we can pull from
//    protected final RandomCollection<Area> potentialAreas;

    // how many seconds to count before switching to the next area
    protected final int warningPhaseSeconds;
    protected final int damagePhaseSeconds;

    // how many seconds for a half heart to be applied
    protected final int secondsForHalfHeart;

    // stores messages for particular actions
    protected final MessageSender phaseStartSender;
    protected final MessageSender updatesSender;
    protected final MessageSender damageSender;

    // Stores how many seconds of damage the player has accumulated
    protected final Map<UUID, PlayerDamagePoints> owedDamage = Maps.newHashMap();

    // Holds the current selected areas we will keep checking
    protected AreaSelection selection;

    // current running phase
    protected Phase currentPhase = Phase.WARNING;

    // current timer for current phase
    protected long countdown;

    protected final List<Long> notificationTicks;
    protected final PlayerStrings _;

    protected MigrationRunnable(
            AreaSelection selection,
            MessageSender phaseStartSender, MessageSender updatesSender, MessageSender damageSender,
            int warningPhaseSeconds, int damagePhaseSeconds,
            int secondsForHalfHeart,
            List<Long> notificationTicks,
            PlayerStrings _)
    {
        this.selection = selection;
        this.phaseStartSender = phaseStartSender;
        this.updatesSender = updatesSender;
        this.damageSender = damageSender;
        this.warningPhaseSeconds = warningPhaseSeconds;
        this.damagePhaseSeconds = damagePhaseSeconds;
        this.secondsForHalfHeart = secondsForHalfHeart;
        this.notificationTicks = notificationTicks;
        this._ = _;

        enterWarningPhase();
    }

    protected void enterWarningPhase() {
        currentPhase = Phase.WARNING;
        countdown = warningPhaseSeconds;

        // randomize the area to warn about
        selection.randomize();

        // warn everyone with a title just the once
        for (Player player : Bukkit.getOnlinePlayers()) {
            Area area = selection.getForUUID(player.getUniqueId());
            String message = _.format("GET TO AREA", area.getAnnounce(), countdown);

            phaseStartSender.sendPlayerMessage(player, message);
        }
    }

    protected void enterDamagingPhase() {
        currentPhase = Phase.DAMAGING;
        countdown = damagePhaseSeconds;

        // tell people with a title just the once
        String message = _.format("DAMAGE PHASE START");
        for (Player player : Bukkit.getOnlinePlayers()) {
            phaseStartSender.sendPlayerMessage(player, message);
        }
    }

    @Override
    public void run() {
        countdown--;

        // update the countdown if it ran out
        if (countdown == 0) {
            if (currentPhase == Phase.WARNING) {
                enterDamagingPhase();
            } else {
                enterWarningPhase();
            }

            // don't do anything else this second
            return;
        }

        boolean announceUpdate = notificationTicks.contains(countdown);

        if (currentPhase == Phase.WARNING) {
            // update people with how long they have to get to the area
            if (announceUpdate) {
                for (Player player : Bukkit.getOnlinePlayers()) {
                    Area area = selection.getForUUID(player.getUniqueId());
                    String message = _.format("GET TO AREA", area.getAnnounce(), countdown);

                    updatesSender.sendPlayerMessage(player, message);
                }
            }

            return;
        }

        // Phase.DAMAGING

        for (Player player : Bukkit.getOnlinePlayers()) {
            Location location = player.getLocation();
            Area area = selection.getForUUID(player.getUniqueId());

            boolean inside = area.inside(location.getX(), location.getZ());

            // if they are not inside the current selection add a point of damage to them
            if (!inside) {
                PlayerDamagePoints damage = owedDamage.get(player.getUniqueId());

                // set if it doesn't exist
                if (damage == null) {
                    damage = new PlayerDamagePoints();
                    owedDamage.put(player.getUniqueId(), damage);
                }

                // add one onto the owed damage and pay any health needed
                GameMode mode = player.getGameMode();
                if (mode == GameMode.SURVIVAL || mode == GameMode.ADVENTURE) {
                    damage.increment();
                    attemptToApplyDamage(player);
                }
            }

            if (announceUpdate) {
                String message = _.format("DAMAGE TICK", area.getAnnounce(), countdown);
                updatesSender.sendPlayerMessage(player, message);
            }
        }
    }

    /**
     * Checks accumulated damage for the player and applies it as required
     *
     * @param player the player to check for
     */
    public void attemptToApplyDamage(Player player) {
        PlayerDamagePoints points = owedDamage.get(player.getUniqueId());

        // player hasn't taken any points yet
        if (points == null) return;

        int times = 0;
        while (points.get() >= secondsForHalfHeart) {
            times++;
            // remove the points and increase the amount of times we need to damage
            points.remove(secondsForHalfHeart);
        }

        // if they owe health
        if (times > 0) {
            player.setHealth(Math.max(0, player.getHealth() - (times)));
            damageSender.sendPlayerMessage(player, _.format("DAMAGE NOTIFICATION", times));
        }
    }
}

File: src/main/java/gg/uhc/migration/MigrationRunnable.kt
package gg.uhc.migration

import com.google.common.collect.Maps
import gg.uhc.migration.messages.MessageSender
import gg.uhc.migration.selection.AreaSelection
import org.bukkit.Bukkit
import org.bukkit.GameMode
import org.bukkit.entity.Player
import org.bukkit.scheduler.BukkitRunnable
import java.util.UUID

class MigrationRunnable (
    // Holds the current selected areas we will keep checking
    private var selection: AreaSelection,
    // stores messages for particular actions
    private val phaseStartSender: MessageSender,
    private val updatesSender: MessageSender,
    private val damageSender: MessageSender,

    // how many seconds to count before switching to the next area
    private val warningPhaseSeconds: Int,
    private val damagePhaseSeconds: Int,

    // how many seconds for a half heart to be applied
    private val secondsForHalfHeart: Int,
    private val notificationTicks: List<Long>,
    private val messages: CustomMessages
) : BukkitRunnable() {
    private enum class Phase {
        WARNING,
        DAMAGING
    }

    // Stores how many seconds of damage the player has accumulated
    private val owedDamage: MutableMap<UUID, Int> = Maps.newHashMap<UUID, Int>()

    // current running phase
    private var currentPhase = Phase.WARNING

    // current timer for current phase
    private var countdown: Long = 0

    init {
        enterWarningPhase()
    }

    private fun enterWarningPhase() {
        currentPhase = Phase.WARNING
        countdown = warningPhaseSeconds.toLong()

        // randomize the area to warn about
        selection.randomize()

        // warn everyone with a title just the once
        Bukkit
            .getOnlinePlayers()
            .map { Pair(it, messages["GET TO AREA", selection.getForUUID(it.uniqueId).announce, countdown]) }
            .forEach { phaseStartSender.sendPlayerMessage(it.first, it.second) }
    }

    private fun enterDamagingPhase() {
        currentPhase = Phase.DAMAGING
        countdown = damagePhaseSeconds.toLong()

        // tell people with a title just the once
        val message = messages["DAMAGE PHASE START"]
        Bukkit
            .getOnlinePlayers()
            .forEach {
                phaseStartSender.sendPlayerMessage(it, message)
            }
    }

    override fun run() {
        countdown--

        // update the countdown if it ran out
        if (countdown == 0L) {
            when (currentPhase) {
                Phase.WARNING -> enterDamagingPhase()
                else -> enterWarningPhase()
            }

            // don't do anything else this second
            return
        }

        val announceUpdate = notificationTicks.contains(countdown)

        if (currentPhase == Phase.WARNING) {
            // update people with how long they have to get to the area
            if (announceUpdate) {
                Bukkit
                    .getOnlinePlayers()
                    .map { Pair(it, messages["GET TO AREA", selection.getForUUID(it.uniqueId), countdown])}
                    .forEach { updatesSender.sendPlayerMessage(it.first, it.second) }
            }

            return
        }

        // Phase.DAMAGING

        Bukkit
            .getOnlinePlayers()
            .map { Pair(it, selection.getForUUID(it.uniqueId)) }
            .forEach { (p, area) ->
                val location = p.location

                val inside = area.areCoordinatesInside(location.x, location.z)

                // if they are not inside the current selection add a point of damage to them
                if (!inside) {

                    // add one onto the owed damage and pay any health needed
                    when (p.gameMode) {
                        GameMode.SURVIVAL, GameMode.ADVENTURE -> {
                            val damage = owedDamage.getOrDefault(p.uniqueId, 0) + 1
                            owedDamage.put(p.uniqueId, damage)
                            attemptToApplyDamage(p)
                        }
                        else -> {}
                    }
                }

                if (announceUpdate) {
                    updatesSender.sendPlayerMessage(p, messages["DAMAGE TICK", area.announce, countdown])
                }
            }
    }

    /**
     * Checks accumulated damage for the player and applies it as required

     * @param player the player to check for
     */
    fun attemptToApplyDamage(player: Player) {
        var points: Int = owedDamage[player.uniqueId] ?: return

        val times = points / secondsForHalfHeart
        points %= secondsForHalfHeart

        owedDamage.put(player.uniqueId, points)

        // if they owe health
        if (times > 0) {
            player.health = Math.max(0.0, player.health - times)
            damageSender.sendPlayerMessage(player, messages["DAMAGE NOTIFICATION", times])
        }
    }
}
